Client/Frontend/Home/BookmarksEditPanel.swift

```diff
@@ -0,0 +1,335 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import UIKit
+import Storage
+import Shared
+import XCGLogger
+
+private let log = Logger.browserLogger
+
+enum DetailPanel: Int {
+    case Title = 0
+    case URL
+    case Folder
+    case _numberOfPanels
+}
+
+struct BookmarksEditPanelUX {
+    private static let LabelLeftRightPadding = 20
+    private static let LabelTopBottomPadding = 10
+    private static let TextFieldTopBottomPadding = 5
+}
+
+class BookmarksEditPanel: SiteTableViewController, HomePanel, UITextFieldDelegate {
+    weak var homePanelDelegate: HomePanelDelegate? = nil
+    var source: BookmarksModel?
+    var parentFolders = [BookmarkFolder]()
+    var bookmarkFolder: BookmarkFolder?
+    var bookmark: BookmarkItem?
+    
+    var titleField: UITextField?
+    var urlField: UITextField?
+    var doneButton: UIButton?
+    
+    lazy var deleteButtonCell = ButtonPanelCell(style: UITableViewCellStyle.Default, reuseIdentifier: "ButtonPanelCell")
+    var deleteButton: UIButton!
+    
+    private let BookmarksEditPanelCellIdentifier = "BookmarksEditPanelCellIdentifier"
+    private let BookmarkFolderEditHeaderViewIdentifier = "BookmarkFolderEditHeaderIdentifier"
+    
+    init() {
+        super.init(nibName: nil, bundle: nil)
+        
+        deleteButton = deleteButtonCell.button
+        deleteButton.setTitle(deleteDeleteButtonLabel, forState: UIControlState.Normal)
+        deleteButton.setTitleColor(UIColor.redColor(), forState: UIControlState.Normal)
+        deleteButton.setTitleColor(UIColor.redColor().colorWithAlphaComponent(0.5), forState: UIControlState.Highlighted)
+        deleteButton.setTitleColor(UIColor.grayColor(), forState: UIControlState.Disabled)
+        deleteButton.addTarget(self, action: #selector(BookmarksEditPanel.deletePressed), forControlEvents: .TouchUpInside)
+        
+        self.tableView.registerClass(BookmarkFolderEditTableViewHeader.self, forHeaderFooterViewReuseIdentifier: BookmarkFolderEditHeaderViewIdentifier)
+        self.tableView.registerClass(BookmarksEditPanelCell.self, forCellReuseIdentifier: BookmarksEditPanelCellIdentifier)
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.tableView.accessibilityIdentifier = "Bookmarks List"
+        self.tableView.backgroundColor = UIColor(red: 240/255, green: 240/255, blue: 240/255, alpha: 1)
+    }
+    
+    func deletePressed() {
+        guard let source = self.source,
+              let bookmark = self.bookmark else {
+                return
+        }
+        
+        guard let factory = source.modelFactory.value.successValue else {
```

Comment:
Avoid calling `value` on instances of `Deferred`. This will force a (usually) asynchronous call to be called synchronously and block the current thread (in this case the main thread) until the execution is complete. In this case, I would chain in the remove call. See https://github.com/mozilla/firefox-ios/blob/master/Client/Frontend/Browser/BrowserViewController.swift#L816 for a similar implementation.